File: src/compiler/frontend/PreScanVisitor.cpp
/*
 * -----------------------------------------------------------------------------
 * PyMCU Compiler (pymcuc)
 * Copyright (C) 2026 Ivan Montiel Cardona and the PyMCU Project Authors
 *
 * SPDX-License-Identifier: MIT
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * -----------------------------------------------------------------------------
 * SAFETY WARNING / HIGH RISK ACTIVITIES:
 * THE SOFTWARE IS NOT DESIGNED, MANUFACTURED, OR INTENDED FOR USE IN HAZARDOUS
 * ENVIRONMENTS REQUIRING FAIL-SAFE PERFORMANCE, SUCH AS IN THE OPERATION OF
 * NUCLEAR FACILITIES, AIRCRAFT NAVIGATION OR COMMUNICATION SYSTEMS, AIR
 * TRAFFIC CONTROL, DIRECT LIFE SUPPORT MACHINES, OR WEAPONS SYSTEMS.
 * -----------------------------------------------------------------------------
 */

#include "PreScanVisitor.h"

#include <iostream>
#include <stdexcept>

#include "../common/Errors.h"
#include "Ast.h"

PreScanVisitor::PreScanVisitor(DeviceConfig& config) : config(config) {}

void PreScanVisitor::scan(const Program& program) const {
  for (const auto& stmt : program.global_statements) {
    visit_statement(stmt.get());
  }
}

void PreScanVisitor::visit_statement(const Statement* stmt) const {
  if (const auto exprStmt = dynamic_cast<const ExprStmt*>(stmt)) {
    if (const auto call = dynamic_cast<const CallExpr*>(exprStmt->expr.get())) {
      if (const auto var =
              dynamic_cast<const VariableExpr*>(call->callee.get())) {
        if (var->name == "device_info") {
          handle_device_info(call);
        }
      }
    }
  }
}

void PreScanVisitor::handle_device_info(const CallExpr* call) const {
  int positional_index = 0;

  for (const auto& arg : call->args) {
    std::string key;
    const Expression* valueExpr = nullptr;

    if (const auto kw = dynamic_cast<const KeywordArgExpr*>(arg.get())) {
      key = kw->key;
      valueExpr = kw->value.get();
    } else {
      switch (positional_index) {
        case 0:
          key = "arch";
          break;
        case 1:
          key = "chip";
          break;
        case 2:
          key = "ram_size";
          break;
        case 3:
          key = "flash_size";
          break;
        case 4:
          key = "eeprom_size";
          break;
        default:
          throw CompilerError("ConfigError",
                              "Too many positional arguments in device_info",
                              call->line, 0);
      }
      valueExpr = arg.get();
      positional_index++;
    }

    if (key == "chip") {
      if (const auto lit = dynamic_cast<const StringLiteral*>(valueExpr)) {
        std::string parsed_chip = lit->value;
        config.detected_chip = parsed_chip;
        config.chip = parsed_chip;

        // Validation: CLI/TOML vs source code chip
        // Source code device_info() takes precedence over CLI --arch,
        // since the code was written for a specific chip.
        if (!config.target_chip.empty() && config.target_chip != parsed_chip) {
          std::cerr << "[Warning] Build specifies '" << config.target_chip
                    << "', but code imports '" << parsed_chip
                    << "'. Using source code chip.\n";
          config.target_chip = parsed_chip;
        }

        // Auto-detect architecture if not already set
        if (config.arch.empty()) {
          if (parsed_chip.find("pic16f1") == 0 && parsed_chip.length() >= 9) {
            config.arch = "pic14e";
          } else {
            config.arch = "pic14";
          }
        }
      } else {
        throw CompilerError("ConfigError", "chip must be a string literal",
                            call->line, 0);
      }
    } else if (key == "arch") {
      if (const auto lit = dynamic_cast<const StringLiteral*>(valueExpr)) {
        config.arch = lit->value;
      } else {
        throw CompilerError("ConfigError", "arch must be a string literal",
                            call->line, 0);
      }
    } else if (key == "ram_size") {
      if (const auto lit = dynamic_cast<const IntegerLiteral*>(valueExpr)) {
        config.ram_size = lit->value;
      }
      // Non-literal (e.g., variable reference) is allowed; value will be 0
    } else if (key == "flash_size") {
      if (const auto lit = dynamic_cast<const IntegerLiteral*>(valueExpr)) {
        config.flash_size = lit->value;
      }
    } else if (key == "eeprom_size") {
      if (const auto lit = dynamic_cast<const IntegerLiteral*>(valueExpr)) {
        config.eeprom_size = lit->value;
      }
    }
  }

  std::cout << "[PreScan] Configured device: " << config.chip
            << " (Arch: " << config.arch << ")"
            << " (RAM: " << config.ram_size << ", Flash: " << config.flash_size
            << ")\n";
}

File: src/compiler/Frontend/PreScanVisitor.cs
/*
 * -----------------------------------------------------------------------------
 * PyMCU Compiler (pymcuc)
 * Copyright (C) 2026 Ivan Montiel Cardona and the PyMCU Project Authors
 *
 * SPDX-License-Identifier: MIT
 *
 * -----------------------------------------------------------------------------
 * SAFETY WARNING / HIGH RISK ACTIVITIES:
 * THE SOFTWARE IS NOT DESIGNED, MANUFACTURED, OR INTENDED FOR USE IN HAZARDOUS
 * ENVIRONMENTS REQUIRING FAIL-SAFE PERFORMANCE, SUCH AS IN THE OPERATION OF
 * NUCLEAR FACILITIES, AIRCRAFT NAVIGATION OR COMMUNICATION SYSTEMS, AIR
 * TRAFFIC CONTROL, DIRECT LIFE SUPPORT MACHINES, OR WEAPONS SYSTEMS.
 * -----------------------------------------------------------------------------
 */

using PyMCU.Common;
using PyMCU.Common.Models;

namespace PyMCU.Frontend;

// Scans the AST for configuration calls like device_info()
// This runs BEFORE type checking or code generation.
public class PreScanVisitor(DeviceConfig config)
{
    // isTargetEstablished = true  → target was set by BootstrapPhase; validate only.
    // isTargetEstablished = false → first device_info() seen establishes the target.
    // Module-level integer constants of the file being scanned, so that
    // `RAM_SIZE = 2048` ... `device_info(ram_size=RAM_SIZE)` resolves. Every chip
    // file in the stdlib is written that way, and until this map existed the sizes
    // were silently dropped: the argument was a name, not an IntegerLiteral, the
    // `is IntegerLiteral` test simply did not match, and DeviceConfig kept its 0.
    private readonly Dictionary<string, int> moduleConstants = new();

    public void Scan(ProgramNode program, bool isTargetEstablished = false)
    {
        // Collected up front, so a constant declared after the device_info() call
        // resolves as readily as one declared before it.
        foreach (var stmt in program.GlobalStatements)
            RecordConstant(stmt);

        foreach (var stmt in program.GlobalStatements)
        {
            VisitStatement(stmt, isTargetEstablished);
        }
    }

    private void RecordConstant(Statement stmt)
    {
        switch (stmt)
        {
            case VarDecl { Init: IntegerLiteral v } d:
                moduleConstants[d.Name] = v.Value;
                break;
            case AnnAssign { Value: IntegerLiteral v } a:
                moduleConstants[a.Target] = v.Value;
                break;
            case AssignStmt { Target: VariableExpr t, Value: IntegerLiteral v }:
                moduleConstants[t.Name] = v.Value;
                break;
        }
    }

    // An integer argument to device_info(): a literal, or a module-level integer
    // constant of the same file. Anything else is refused rather than dropped --
    // a size that quietly stays 0 is exactly the failure this plumbing exists to end.
    private int ResolveSize(string key, Expression valueExpr, int line)
    {
        if (valueExpr is IntegerLiteral lit) return lit.Value;

        if (valueExpr is VariableExpr name && moduleConstants.TryGetValue(name.Name, out int v))
            return v;

        throw new CompilerError("ConfigError",
            $"device_info({key}=...) must be an integer literal or a module-level integer "
            + "constant of this file; the backends read this number and cannot be given "
            + "an expression they must evaluate.",
            line, 0);
    }

    private void VisitStatement(Statement stmt, bool isTargetEstablished)
    {
        if (stmt is ExprStmt { Expr: CallExpr { Callee: VariableExpr { Name: "device_info" } } call })
        {
            HandleDeviceInfo(call, isTargetEstablished);
        }
    }

    private void HandleDeviceInfo(CallExpr call, bool isTargetEstablished)
    {
        int positionalIndex = 0;

        foreach (var arg in call.Args)
        {
            string key = "";
            Expression valueExpr;

            if (arg is KeywordArgExpr kw)
            {
                key = kw.Key;
                valueExpr = kw.Value;
            }
            else
            {
                switch (positionalIndex)
                {
                    case 0: key = "arch"; break;
                    case 1: key = "chip"; break;
                    case 2: key = "ram_size"; break;
                    case 3: key = "flash_size"; break;
                    case 4: key = "eeprom_size"; break;
                    default:
                        throw new CompilerError("ConfigError", "Too many positional arguments in device_info",
                            call.Line, 0);
                }

                valueExpr = arg;
                positionalIndex++;
            }

            if (key == "chip")
            {
                if (valueExpr is StringLiteral lit)
                {
                    string parsedChip = lit.Value;

                    if (isTargetEstablished)
                    {
                        // Target is locked. Validate arch-family compatibility.
                        // Same chip family (e.g. atmega328 in an atmega328p project): warning only.
                        // Different family (e.g. pic18 chip in an avr project): fatal error.
                        if (!ArchFamilyResolver.SameFamily(config.Arch, parsedChip))
                        {
                            var importedFamily = ArchFamilyResolver.Resolve(parsedChip);
                            var targetFamily   = ArchFamilyResolver.Resolve(config.Arch);
                            throw new CompilerError("TargetMismatch",
                                $"Cross-architecture import detected: chip file declares '{parsedChip}' " +
                                $"(family={importedFamily}) but the build target is '{config.TargetChip}' " +
                                $"(family={targetFamily}). Cannot mix chip definitions from different " +
                                $"architecture families. Check your imports or your [tool.pymcu] target.",
                                call.Line, 0);
                        }

                        if (config.TargetChip != parsedChip)
                        {
                            Logger.Warning("PreScan",
                                $"Chip file declares '{parsedChip}' but build target is '{config.TargetChip}'. " +
                                $"Same architecture family — proceeding with build target.");
                        }
                        // Do NOT overwrite config.Chip or config.TargetChip.
                    }
                    else
                    {
                        // Bootstrap mode: first device_info() establishes the target.
                        config.DetectedChip = parsedChip;
                        config.Chip = parsedChip;

                        if (!string.IsNullOrEmpty(config.TargetChip) && config.TargetChip != parsedChip)
                        {
                            Logger.Warning("PreScan",
                                $"Target '{config.TargetChip}' (from build config) differs from chip file '{parsedChip}'. " +
                                $"Using build target.");
                        }
                    }
                }
                else
                {
                    throw new CompilerError("ConfigError", "chip must be a string literal", call.Line, 0);
                }
            }
            else if (key == "arch")
            {
                if (valueExpr is StringLiteral lit)
                {
                    if (!isTargetEstablished)
                        config.Arch = lit.Value;
                    // When established: arch= in a non-bootstrap chip file is ignored;
                    // family validation already happened via the chip= key above.
                }
                else
                {
                    throw new CompilerError("ConfigError", "arch must be a string literal", call.Line, 0);
                }
            }
            else if (key == "ram_size")
            {
                if (!isTargetEstablished)
                    config.RamSize = ResolveSize(key, valueExpr, call.Line);
            }
            else if (key == "flash_size")
            {
                if (!isTargetEstablished)
                    config.FlashSize = ResolveSize(key, valueExpr, call.Line);
            }
            else if (key == "eeprom_size")
            {
                if (!isTargetEstablished)
                    config.EepromSize = ResolveSize(key, valueExpr, call.Line);
            }
        }

        Logger.Verbose("PreScan",
            $"Configured device: {config.Chip} (Arch: {config.Arch}) (RAM: {config.RamSize}, Flash: {config.FlashSize})");
    }
}